File: backend/core/bankroll/kelly.py
"""Kelly Criterion + Expected Value + Verdict logic.

Pure functions. Math validated by unit tests with known edge cases.

References:
    Kelly, J. L. (1956). "A New Interpretation of Information Rate".
    Cap (max 25%) is convention to avoid ruin from variance.
"""

from core.types import (
    EV,
    Apostar,
    Esperar,
    KellyFraction,
    NoApostar,
    Odds,
    Probability,
    Stake,
    Verdict,
)


# Decision thresholds (tunable; documented in plan).
EV_APOSTAR_MIN = 0.05      # Need >5% expected profit to recommend bet
KELLY_APOSTAR_MIN = 0.02   # AND at least 2% of bankroll
KELLY_CAP = 0.25           # Never bet more than 25% even if Kelly says more


def compute_ev(prob: Probability, odds: Odds, stake: Stake) -> EV:
    """Expected Value of the bet.

        EV = p * (odds - 1) * stake  -  (1 - p) * stake

    Positive EV → long-run profitable. Negative EV → don't bet.
    """
    gain_if_win = (odds - 1.0) * stake
    loss_if_lose = stake
    ev = prob * gain_if_win - (1.0 - prob) * loss_if_lose
    return EV(ev)


def kelly_fraction(
    prob: Probability, odds: Odds, cap: float = KELLY_CAP,
) -> KellyFraction:
    """Kelly Criterion: fraction of bankroll to bet for optimal log-growth.

        f* = (b * p - q) / b   where b = odds - 1, q = 1 - p

    Returns 0 when:
    - No edge (negative Kelly)
    - Odds <= 1 (no payout above stake)
    Caps at `cap` (default 0.25) to limit drawdown from variance.
    """
    b = odds - 1.0
    if b <= 0 or prob <= 0:
        return KellyFraction(0.0)

    q = 1.0 - prob
    f_star = (b * prob - q) / b

    if f_star <= 0:
        return KellyFraction(0.0)

    return KellyFraction(min(f_star, cap))


def verdict_from_ev_kelly(
    ev: EV, kelly: KellyFraction, reason_context: str = "",
) -> Verdict:
    """Decide based on EV and Kelly thresholds.

    - APOSTAR: ev > EV_APOSTAR_MIN AND kelly > KELLY_APOSTAR_MIN
    - ESPERAR: any positive signal but below the bet threshold
    - NO_APOSTAR: zero/negative signal
    """
    if ev > EV_APOSTAR_MIN and kelly > KELLY_APOSTAR_MIN:
        return Apostar(reason=f"EV={ev:.2f}, Kelly={kelly:.2%}. {reason_context}".strip())

    if ev > 0 or kelly > 0:
        return Esperar(
            reason=f"Marginal: EV={ev:.2f}, Kelly={kelly:.2%}. {reason_context}".strip()
        )

    return NoApostar(reason=f"No edge: EV={ev:.2f}. {reason_context}".strip())

File: mobile/core/kelly.js
// Port of backend/core/bankroll/kelly.py — same formulas + thresholds
// + cap so JS predictions yield identical EV / Kelly / verdict.

export const EV_APOSTAR_MIN = 0.05;      // EV must exceed this to recommend
export const KELLY_APOSTAR_MIN = 0.02;   // AND at least 2% bankroll
export const KELLY_CAP = 0.25;           // never bet > 25%

// EV = p * (odds - 1) * stake  -  (1 - p) * stake.
// Positive → long-run profitable. Stake is the amount at risk.
export function computeEv(prob, odds, stake) {
  const gainIfWin = (odds - 1.0) * stake;
  const lossIfLose = stake;
  return prob * gainIfWin - (1.0 - prob) * lossIfLose;
}

// Kelly fraction f* = (b*p - q) / b  where b = odds-1, q = 1-p.
// Returns 0 when there's no edge or odds <= 1. Capped at KELLY_CAP.
export function kellyFraction(prob, odds, cap = KELLY_CAP) {
  const b = odds - 1.0;
  if (b <= 0 || prob <= 0) return 0.0;
  const q = 1.0 - prob;
  const fStar = (b * prob - q) / b;
  if (fStar <= 0) return 0.0;
  return Math.min(fStar, cap);
}

// Decision tree mirrors Python:
//   ev > 0.05 AND kelly > 0.02  → 'apostar'
//   ev > 0  OR  kelly > 0       → 'esperar'
//   else                        → 'no_apostar'
export function verdictFromEvKelly(ev, kelly, reasonContext = '') {
  const fmt = (n) => n.toFixed(2);
  const pct = (n) => `${(n * 100).toFixed(2)}%`;
  const ctx = reasonContext ? ` ${reasonContext}` : '';

  if (ev > EV_APOSTAR_MIN && kelly > KELLY_APOSTAR_MIN) {
    return { verdict: 'apostar', reason: `EV=${fmt(ev)}, Kelly=${pct(kelly)}.${ctx}` };
  }
  if (ev > 0 || kelly > 0) {
    return { verdict: 'esperar', reason: `Marginal: EV=${fmt(ev)}, Kelly=${pct(kelly)}.${ctx}` };
  }
  return { verdict: 'no_apostar', reason: `No edge: EV=${fmt(ev)}.${ctx}` };
}
